coverage/src/main/java/com/github/lizardev/xquery/saxon/coverage/trace/CoverageInstructionInjector.java

```diff
@@ -11,27 +11,29 @@
 import net.sf.saxon.om.StructuredQName;
 import net.sf.saxon.query.QueryModule;
 
-import java.util.Map;
-import java.util.WeakHashMap;
+import java.util.concurrent.ConcurrentMap;
 
 import static com.github.lizardev.xquery.saxon.coverage.collect.ModuleId.uniqueModuleId;
+import static com.google.common.cache.CacheBuilder.newBuilder;
 
 public class CoverageInstructionInjector extends TraceCodeInjectorComponent {
 
-    private final Map<QueryModule, ModuleId> moduleIds = new WeakHashMap<>();
+    private final ConcurrentMap<QueryModule, ModuleId> moduleIds;
     private final CoverageInstructionEventHandler eventHandler;
 
     public CoverageInstructionInjector(CoverageInstructionEventHandler eventHandler) {
         this.eventHandler = eventHandler;
+        Cache<QueryModule, ModuleId> cache = newBuilder().weakKeys().weakValues().build();
```

Comment:
btw w/o improving concurrency in CoverageInstructionEventHandler it probably will not improve perf.